### PR TITLE
Xwords: highlight intersecting entries

### PR DESCRIPTION
**Pull request by OliverJAsh**
As per the designs.

This is my first dive into crosswords. There could be existing functions I should be re-using, or I may have put things in the wrong place. Please shout if that's the case.

Merging into @sndrs’s `xwords` branch because I'm based off that and it's not yet in master.

image

### CODE REVIEW

File: static/src/javascripts/es6/projects/common/modules/crosswords/main.js [hunk 1]
```diff
@@ -368,6 +368,23 @@ var Crossword = React.createClass({
         let focussed = this.clueInFocus(),
             isHighlighted = (x, y) => focussed ? helpers.entryHasCell(focussed, x, y) : false;
 
+        // Deep equal version of _.intersection
+        let findIntersectingCells = (array1, array2) =>
```
Comment (robertberry) [inline static/src/javascripts/es6/projects/common/modules/crosswords/main.js]: I think we should probably prefer `const` as the default way of declaring things. That way when you see `let` you know that it's specifically something that's being mutated at some point.

File: static/src/javascripts/es6/projects/common/modules/crosswords/main.js [hunk 2]
```diff
@@ -368,6 +368,23 @@ var Crossword = React.createClass({
...
+            let intersecting = findIntersectingCells(cells, focussedCells);
+            return !! intersecting.length;
+        };
+
+        let otherEntries = _.difference(this.props.data.entries, focussed ? [focussed] : []);
+        let intersectingEntries = otherEntries.filter(entryHasIntersectingCell);
+
+        let cellIntersectsFocussedEntry = (x, y) =>
+            intersectingEntries.some(entry => helpers.entryHasCell(entry, x, y));
+
```
Comment (robertberry) [inline static/src/javascripts/es6/projects/common/modules/crosswords/main.js]: Computationally this is a bit expensive.

If you have n clues, of average m length, you'll be doing nm² operations to get the intersecting clues.

So for this crossword, for example - http://m.code.dev-theguardian.com/crosswords/quick/13864 - you'd be doing 26 * 6 * 6 (936) passes each time.

I doubt that's going to cause any issues for any of the devices accessing the crosswords, but if we do see problems, we could look at caching the intersections on start up, as it's something you can just pre-compute. But don't do that unless we actually see it as a problem at some point, no point in prematurely optimising!

**Comment by robertberry** [comment]
Looks good! :+1:

File: static/src/stylesheets/module/crosswords/_grid.scss [hunk 3]
```diff
@@ -8,6 +8,10 @@
     fill: #ffffff;
 }
 
+.crossword__cell--intersecting {
+    fill: hsl(0, 0%, 80%);
```
Comment (sndrs) [inline static/src/stylesheets/module/crosswords/_grid.scss]: should really be one of our neutral colours: http://guardian.github.io/pasteup-palette/demo/

**Pull request merged**